### PR TITLE
bpo-43153: Don't mask `PermissionError` with `NotADirectoryError` during tempdirectory cleanup

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
Co-Authored-By: andrei kulakov <andrei.avk@gmail.com>

https://bugs.python.org/issue43153

#87319

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
I've tested locally and it works on Windows.

File: Lib/tempfile.py
Comment (vstinner) [inline Lib/tempfile.py]: I don't understand the comment. There is now an explicit `except IsADirectoryError:`. We are handling PermissionError. I don't understand the relationship between PermissionError and IsADirectoryError.

Comment (Fidget-Spinner) [reply Lib/tempfile.py]: My comment needs improving then :). The current bug is that on Windows, calling `_rmtree` when there's a `PermissionError` raises `IsADirectoryError`, which masks the original `PermissionError`. I was trying to convey that but seems that the comment wasn't clear.

File: Lib/tempfile.py
Comment (vstinner) [inline Lib/tempfile.py]: is IsADirectoryError only raised if it's a file?

Comment (Fidget-Spinner) [reply Lib/tempfile.py]: I'm not sure, but it should be ok to raise `IsADirectoryError` for non-file stuff right? If we can't clear a folder due to both `PermissionError` and `IsADirectoryError`, I think it's okay to raise either exception. It's not too clear which one should take precedence.

In the bpo, @akulakov suggested only re-raising for files since `IsADirectoryError` is definitely wrong in those cases.

Comment (vstinner) [reply Lib/tempfile.py]: The issue is about **NotA**DirectoryError, but your change and your comment only mention **IsA**DirectoryError. I'm still confused.

Comment (Fidget-Spinner) [reply Lib/tempfile.py]: OMG, I feel like an idiot now 😆 . Thanks for pointing that out. Yep the comment is wrong, it should be about `NotADirectoryError`.

File: Lib/tempfile.py
Comment (vstinner) [inline Lib/tempfile.py]: maybe test "not isdir()"?

File: Lib/tempfile.py
Comment (akulakov) [inline Lib/tempfile.py]: I think this would also happen on other OSes, not just Windows. The bug can only be reproduced on windows because PermissionError in the bug was caused by an open file, but PermissionError in theory can be caused by actual permission / ownership settings. It shouldn't matter much because this is a temp dir (and files in it) created by current user, but this comment can be misunderstood, so it's better to drop ref. to Windows here.

It may be left in the `if` case below because it can avoid the `isdir()` check.

It would be good to also add "Issue #43153" to the comment.

**Review by akulakov — APPROVED** [review]
Approving with a minor request to update a comment

**Comment by akulakov** [comment]
It might also be good to run this on all windows build bots to confirm that unit test doesn't fail on any windows version we support. The code itself should be fine but unit test may fail if some windows version doesn't raise permission error in this case.

File: Lib/test/test_tempfile.py
Comment (akulakov) [inline Lib/test/test_tempfile.py:1646]: Please add # Issue #43153 here too.

Comment (akulakov) [reply Lib/test/test_tempfile.py:1646]: Actually issue can be found by looking it up via github PR, I've seen these comments on unit tests, but maybe they're not necessary.

**Comment by akulakov** [comment]
Please also update the PR title IsAdir.. => NotADir...

File: Lib/tempfile.py
Comment (eryksun) [reply Lib/tempfile.py]: Note that calling `unlink()` on a directory in Windows always raises `PermissionError`. Windows maps the NT status code `STATUS_FILE_IS_A_DIRECTORY` to the WinAPI error code `ERROR_ACCESS_DENIED`.

Comment (Fidget-Spinner) [reply Lib/tempfile.py]: > I think this would also happen on other OSes, not just Windows. The bug can only be reproduced on windows

I'm confused, should we continue the `_os.name == 'nt'` check then? I'm not worried about the additional performance hit from `isdir` (though I would like to avoid it). Correctness is more important. If this *does* happen on other OSes, then we should remove that OS check.

File: Lib/test/test_tempfile.py
Comment (Fidget-Spinner) [reply Lib/test/test_tempfile.py:1646]: Sure!

I've seen many arguments for and against that. Some people say we don't need it because we have git history, while others say CPython has changed version control system 3 times :). Personally, I add them if a piece of code looks awfully strange to save the next reader some time, and this PR definitely qualifies.

I wonder how things like `#Issue #43153` will hold up too, considering we're moving to GH issues soon, and the GH issue numbers will differ from the BPO ones. Hopefully `bpo-43153` is enough.

File: Lib/tempfile.py
Comment (akulakov) [reply Lib/tempfile.py]: We can remove the os check. It can happen on other OSes. With the check removed, the logic becomes "if we are not permitted to remove the file, return or reraise", which makes perfect sense.

**Comment by Fidget-Spinner** [comment]
Azure pipelines failure is unrelated and fixed by GH-29963.

**Comment by akulakov** [comment]
We can probably just keep the unit test for windows and skip it on other OSes. Other OSes would need different ways of causing a PermissionError, so it's not worth the effort.

**Comment by Fidget-Spinner** [comment]
> We can probably just keep the unit test for windows and skip it on other OSes. Other OSes would need different ways of causing a PermissionError, so it's not worth the effort. [Edit: for example in my recent testing, on MacOS it's not enough to have a file with no permissions, it is the containing dir that has to lack permissions for the file to raise PermissionError).

Yep, I noticed the other tests are starting to fail. Seems like it's not possible to repro this on the other OSes. I'll make sure to test with buildbots before we merge this PR (and once that docfix PR lands).

BTW, thanks for guiding me throughout this PR :). I'm admittedly not familiar with this part of the stdlib. Thanks to Victor and Eryk too!

**Comment by akulakov** [comment]
> BTW, thanks for guiding me throughout this PR :). I'm admittedly not familiar with this part of the stdlib.

Glad I could help! I haven't done that much PR reviews yet so it was great to get that experience too, especially with the logic I was already a bit familiar with. Please add me as a reviewer in the future on similar PRs!

File: Lib/tempfile.py
Comment (serhiy-storchaka) [inline Lib/tempfile.py]: Move this comment below and merge it with the following comment.

The PermissionError handler was originally added for FreeBSD, but it seems that it is raised on Windows too.

File: Lib/tempfile.py
Comment (serhiy-storchaka) [inline Lib/tempfile.py]: `os.path.isdir()` returns True for a symlink to directory. Either check `os.path.islink()` before `os.path.isdir()` or use `os.stat()` explicitly.

Comment (eryksun) [reply Lib/tempfile.py]: Failing to delete a symlink should be rare in a temp directory. But if it's worth handling here, then it should be consistent with `shutil._rmtree_isdir()`. Unfortunately the latter is written for `os.DirEntry` instances. Here's an updated implementation for `shutil` that can also check a path. I factored out the check for a mount point into a common `_st_ismount(st)` function, which avoids duplicated code and, I think, makes it easier to read.

```python
if not hasattr(os.stat_result, 'st_reparse_tag'):
    def _rmtree_islink(path):
        return os.path.islink(path)

    def _rmtree_isdir(path):
        try:
            if isinstance(path, os.DirEntry):
                return entry.is_dir(follow_symlinks=False)
            return stat.S_ISDIR(os.lstat(path).st_mode)
        except OSError:
            return False
else:
    # In general mount points in Windows are directories, but os.unlink()
    # supports them, so rmtree() handles them as symlinks.
    def _st_ismount(st):
        return (st.st_file_attributes & stat.FILE_ATTRIBUTE_REPARSE_POINT
                and st.st_reparse_tag == stat.IO_REPARSE_TAG_MOUNT_POINT)

    def _rmtree_islink(path):
        try:
            if isinstance(path, os.DirEntry):
                st = entry.stat(follow_symlinks=False)
            else:
                st = os.lstat(path)
            return stat.S_ISLNK(st.st_mode) or _st_ismount(st)
        except OSError:
            return False

    def _rmtree_isdir(path):
        try:
            if isinstance(path, os.DirEntry):
                st = entry.stat(follow_symlinks=False)
            else:
                st = os.lstat(path)
            return stat.S_ISDIR(st.st_mode) and not _st_ismount(st)
        except OSError:
            return False
```

File: Lib/tempfile.py
Comment (eryksun) [inline Lib/tempfile.py]: Note that this check is inconsistent with `shutil.rmtree()`. `os.unlink()` deletes a mount point in Windows, but otherwise a mount point is handled as a directory,  i.e. `islink()` will be false and `isdir()` will be true. However, they're special cased by `shutil._rmtree_isdir()` and `shutil._rmtree_islink()`, and thus `shutil.rmtree()` will immediately fail with `OSError("Cannot call rmtree on a symbolic link")`, as would happen for a regular symbolic link, which is exactly what this check is supposed to prevent.

Comment (Fidget-Spinner) [reply Lib/tempfile.py]: To clarify, so are you suggesting we:
1. Modify shutil._rmtree_isdir with your suggested changes here https://github.com/python/cpython/pull/29940#discussion_r764529814.
2. Use shutil._rmtree_isdir for a more consistent check instead of our current method?

Comment (eryksun) [reply Lib/tempfile.py]: I know it's a bad practice to rely on a private function from another module. But duplicating the code would be a maintenance burden, and it's not useful enough to expose in `os.path`. Maybe `_rmtree_isdir()` and `_rmtree_islink()` should be refactored into a common private module. I'm curious to know what @serhiy-storchaka thinks should be done about the consistency problem between these functions and `os.path.islink()` and `os.path.isdir()`.

Comment (Fidget-Spinner) [reply Lib/tempfile.py]: I'm perfectly fine with using private functions from another module as long as it's within the stdlib itself and not exposed somewhere :). Thank you for the clarification, I wanted to summarize your suggestions since I wasn't sure I understood you correctly.

Comment (akulakov) [reply Lib/tempfile.py]: I think it should be okay to use shutil._rmtree_isdir() here because we're asking shutil if it's valid to use rmtree on a path, and then telling shutil to use shutil._rmtree(). So these two steps of logic stay completely internal to shutil even though used outside of it. And it's also consistent with current logic of using internal shutil._rmtree().

Comment (serhiy-storchaka) [reply Lib/tempfile.py]: `_rmtree_isdir()` and `_rmtree_islink()` was changed between versions, and this fix should be applied to multiple Python versions. So it is better to inline them here. If possible, we should use public API (add it if it does not exist).

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Unfortunately, I failed to create tests for junction points (which would fail without special handling of junction points in tempfile). But the code looks right.

File: Lib/tempfile.py
Comment (eryksun) [inline Lib/tempfile.py:918]: Checking `S_ISLINK(mode)` is redundant here since the code also checks `not S_ISDIR(mode)`. The latter is always true for a symlink, as required by POSIX and as supported by Python on Windows. On the other hand, a junction (mount point) is always a directory on Windows.

Python 3.12+ has an optimized `os.path.isjunction()` function. It's always false on POSIX since POSIX doesn't have a junction file type. With the use of `isjunction()`, this check could be more simply written as follows:

```python
if not _os.path.isdir(path) or _os.path.isjunction(path):
    if ignore_errors:
        return
    raise
cls._rmtree(path, ignore_errors=ignore_errors)
```

Note that `os.unlink()` is implemented on Windows to delete a symlink or junction regardless of whether the target is a file or a directory. It should only fail with a permission error if the reparse point has the readonly file attribute set or if the effective security of the reparse point doesn't grant delete access to the current user.

Comment (serhiy-storchaka) [reply Lib/tempfile.py:918]: Thank you for your suggestion. Done in #112791.

Comment (proc-2) [inline Lib/tempfile.py:918]: Why not go with an "ask-for-forgiveness-rather-than-permission" type of solution, e.g.:

```py
try:
    _os.unlink(path)
except IsADirectoryError:
    cls._rmtree(path, ignore_errors=ignore_errors)
except PermissionError as pe:
    # PermissionError is raised on FreeBSD for directories
    # and by Windows on lock files used by other processes
    try:
        cls._rmtree(path, ignore_errors=ignore_errors)
    except NotADirectoryError:
        # NOTE: This is raised if PermissionError did not
        # correspond to a IsADirectoryError, e.g. on
        # Windows.
        if not ignore_errors:
            raise pe
```

Comment (serhiy-storchaka) [reply Lib/tempfile.py:918]: `rmtree()` is a complex function, and we cannot be sure whether NotADirectoryError was raised by the toplevel `scandir()` or somewhere deeper in the tree.

Comment (proc-2) [reply Lib/tempfile.py:918]: Mmmh...

Comment (proc-2) [reply Lib/tempfile.py:918]: Seems like a responsibility inversion to me. Sure there could be a bug deep down in `_rmtree` that ends up bubbling an unrelated `NotADirectoryError`, but that would technically be a bug in `_rmtree`. Maybe there are valid cases where such an unrelated error would be raised?

Also I see the diff I commented on is not the final version so I'll look at the final version.

Comment (serhiy-storchaka) [reply Lib/tempfile.py:918]: Not necessary a bug. It may be a race condition. Although the LBYL approach is also prone to race conditions, I think it has less chance to override error with a wrong exception.

I myself prefer the EAFP approach, but I do not think that it has advantages in this case.

**Pull request merged**
